Handle new data correctly and extend functionality of MMM posterior predictive methods

wd60622 opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-labs/pymc-marketing/releases -->

## Description
<!--- Describe your changes in detail -->

Redo of the support for new mmm data

This brings in the feedback about 

- having the `include_last_observations` for all predict methods with `X_pred`
- Support for original scale of target with the `original_scale` keyword that can be passed to all predict methods

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #472 #450 #400 #396 
- [ ] Related to #444 #392 

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Modules affected
<!--- Please list the modules that are affected by this PR by typing an `x` in the boxes below: -->
- [x] MMM
- [ ] CLV
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pymc-marketing start -->
----
📚 Documentation preview 📚: https://pymc-marketing--482.org.readthedocs.build/en/482/

<!-- readthedocs-preview pymc-marketing end -->

wd60622 commented on tests/mmm/test_delayed_saturated_mmm.py:
```diff
@@ -579,6 +610,63 @@ def test_model_config(
+        mmm = request.getfixturevalue(model_name)
+        X_pred = generate_data(new_dates)
+
+        pp_without = mmm.predict_posterior(
```
@juanitorduz 
I can still break these down a bit more
I'm trying to get a handful of the edge cases via the parametrize tests

Are you thinking one for each predict method?

juanitorduz replied:
Yes, ideally test should be atomic (test one thing) and have all assert statements at the end. Not a strong opinion, more.of a suggestion :)

wd60622 replied:
Fair enough. I'm pulling out the date ranges and separating the tests

wd60622 commented:
Introducing he new `predict_posterior` call on mmm model

<img width="1279" alt="Screenshot 2024-01-11 at 20 41 28" src="https://github.com/pymc-labs/pymc-marketing/assets/57733339/ad7b2c7f-fc5a-435c-94d5-5fc7be693aea">

ricardoV94 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -83,7 +84,7 @@ def default_sampler_config(self) -> Dict:
     def output_var(self):
         """Defines target variable for the model"""
-        return "y"
+        return "likelihood"
```
I think y is a better good name. likelihood is about the kind of variable it is. That would be like renaming betas to "prior"

wd60622 replied:
I change in the model definition then!

ricardoV94 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +847,80 @@ def plot_channel_contributions_grid(
+            attrs = posterior_predictive_samples.attrs
+
+            if combined:
+                posterior_predictive_samples = xr.apply_ufunc(
```
This kind of complexity/gotchas go away if you make the transforms part of the model itself (but for that you need to write them in PyTensor)

That's why I pushed for ir before. Out of the scope of this PR

wd60622 replied:
I know we've had the discussion when talking about #407. Maybe it is easier to start with the target as a pytensor transformation. 
Any change back to original scale (a few spots) would have to be changed

ricardoV94 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +847,80 @@ def plot_channel_contributions_grid(
+        with self.model:  # sample with new input data
+            post_pred = pm.sample_posterior_predictive(self.idata, **kwargs)
+            if extend_idata:
+                self.idata.extend(post_pred)  # type: ignore
```
You need to define join otherwise calling this method twice will ignore the second run

See https://github.com/pymc-devs/pymc-experimental/pull/249

ricardoV94 replied:
This applies to prior and fit as well, but maybe it's a good time as any to fix it as it's a pretty serious one.

I had a PR on this but haven't come back to it

wd60622 replied:
Sounds good. This is in the method override. Shall I add to the model_builder method as well?

ricardoV94 replied:
Yup. Feel free to reuse as much as you can from the PR of the pymc-experimental ModelBuilder (thinking about tests as well). Just add the author as a co-author if you do that.

wd60622 replied:
Sounds good. I will dive into it more later

wd60622 replied:
I've just added a check that the idata will be the first sample from two method calls to keep it simple. Hope that's what you expected

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +847,80 @@ def plot_channel_contributions_grid(
+        Parameters
+        ---------
+        X_pred : array, shape (n_pred, n_features)
+            The input data used for prediction using prior distribution..
```
Remove last (repeated point)

wd60622 replied:
done

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +847,80 @@ def plot_channel_contributions_grid(
+            Defaults to False.
+        combined: Combine chain and draw dims into sample. Won't work if a dim named sample already exists.
+            Defaults to True.
+        include_last_observations: Boolean determining whether to include the last observations of the training data.
```
Maybe add a reason why this is relevant. I mean, a  comment that this is relevant for the adstock computation.

wd60622 replied:
done

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +847,80 @@ def plot_channel_contributions_grid(
+        """
+        if include_last_observations:
+            X_pred = pd.concat(
+                [self.X.iloc[-self.adstock_max_lag :, :], X_pred], axis=0
```
Shall we sort by date just in case?

wd60622 replied:
done

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +847,80 @@ def plot_channel_contributions_grid(
+        combined,
+        include_last_observations: bool = False,
+        original_scale: bool = True,
+        **kwargs,
```
Maybe call this `sample_posterior_predictive_kwargs` (or something more informative?) To indicate where these additional parameters will be used.

juanitorduz replied:
Or as below a better description: "Additional arguments to pass to sample_posterior_predictive method"

wd60622 replied:
done. This isn't changed for the original model_builder methods however. Would you want that to change in this PR?

juanitorduz replied:
Ok I think we can leave it as it is in the docstrings for now. Let's wait for feedback!

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +847,80 @@ def plot_channel_contributions_grid(
+                    posterior_predictive_samples,
+                )
+            else:
+                posterior_predictive_samples = xr.apply_ufunc(
+                    self.get_target_transformer().inverse_transform,
+                    posterior_predictive_samples.expand_dims(dim={"_": 1}, axis=1),
+                    input_core_dims=[["date", "_"]],
+                    output_core_dims=[["date", "_"]],
+                    vectorize=True,
+                ).squeeze(dim="_")
```
I understand this because I wrote it but it might be confusing. So you could write a little helper function to wrap this logic and add a unit-test (just a suggestion)

wd60622 replied:
Will do!

juanitorduz reviewed (changes requested):
This looks super nice! Thanks! I just added some comments about the docs and other minor things.

juanitorduz commented:
@wd60622 After we merge this one, we need to add (or create) an example to the docs (let's create an issue).  This one will be a game-changer feature 🚀

ricardoV94 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +847,80 @@ def plot_channel_contributions_grid(
+                [self.X.iloc[-self.adstock_max_lag :, :], X_pred], axis=0
+            )
+
+        self._data_setter(X_pred)
```
I wonder if we should reset the data changes after sampling

wd60622 replied:
I'm hear this as setting the MutableData back to the X and y that it is fit on. i.e. self.X and self.y
Is that correct?
I think that would be good to preserve

wd60622 replied:
Added a `_reset_model_data` method to set to the original and it is called after the sample_posterior_predictive call

ricardoV94 replied:
On a second thought I see that all sampling methods need X as input, so I think we should revert this change. Otherwise we should reset in all the other methods. Just in this one doesn't make sense.

Apologies, I find the ModelBuilder API pretty confusing.

ricardoV94 commented on tests/mmm/test_delayed_saturated_mmm.py:
```diff
@@ -580,6 +609,104 @@ def test_model_config(
+    X_pred = generate_data(new_dates)
+
+    posterior_predictive_mean = mmm.predict(X_pred=X_pred)
+    assert posterior_predictive_mean.shape[0] == new_dates.size
```
I would like to test something about the output values, not just their shapes

wd60622 replied:
added some checks based on the original scale and the original y for context

ricardoV94 commented on tests/mmm/test_delayed_saturated_mmm.py:
```diff
@@ -580,6 +609,104 @@ def test_model_config(
+    mmm = request.getfixturevalue(model_name)
+    X_pred = generate_data(new_dates)
+
+    posterior_predictive = mmm.sample_posterior_predictive(
```
What's the difference between the `predict` and `sample_posterior_predictive` MMM methods?

wd60622 replied:
predict is the mean of the sample_posterior_predictive by obs

ricardoV94 commented on tests/model_builder/test_model_builder.py:
```diff
@@ -271,3 +274,27 @@ def test_id():
+    assert first_sample != second_sample
+
+    sample = getattr(fitted_model_instance.idata, name)
+    assert sample == first_sample
```
Shouldn't it override the first with the second when `extend_idata` is True? 

Also we should parametrize this test to test both `extend_idata` cases

ricardoV94 replied:
Also I think these changes also apply to `fit`?

wd60622 replied:
This is the consequent from join="left"
I would think if you call the method again, you'd want to replace or be notified that it is being replaced. 

I think this is out of this PR's scope though\

ricardoV94 replied:
We should join right, otherwise extend_inferencedata=True makes little sense?

wd60622 replied:
That's what I meant to say. That's how I've change it. join="right"

ricardoV94 replied:
This test is still weird.

I expect the assert to be `assert sample == second_sample`.

The fact it is not failing tells me that assert is too premissive. I guess inferencedata equality is just based on groups or something. We need to check some actual values

wd60622 replied:
Okay. I will take a look later

wd60622 commented:
Thank you guys for the review! I will have some time tomorrow to make some changes

juanitorduz reviewed (approved):
This is amazing @wd60622 ! From my side, I just wanna push this out and get feedback. My sleep deprivation (👶 ) might make me miss some details so if @ricardoV94 is happy I  suggest we merge and release 0.4 🚀

ricardoV94 commented on tests/mmm/test_delayed_saturated_mmm.py:
```diff
@@ -580,13 +609,126 @@ def test_model_config(
 
 def test_get_invalid_distribution(mmm):
-    with pytest.raises(ValueError, match="does not exist in PyMC"):
+    with pytest.raises(ValueError, match="does not exist in P"):
```
Why this change?

wd60622 replied:
Not sure. I will change back

wd60622 commented on pymc_marketing/model_builder.py:
```diff
@@ -582,7 +582,7 @@ def sample_prior_predictive(
                 if extend_idata:
                     if self.idata is not None:
-                        self.idata.extend(prior_pred)
+                        self.idata.extend(prior_pred, join="right")
```
@ricardoV94 the join kwarg was added. Maybe you are looking at old commit

wd60622 commented on pymc_marketing/model_builder.py:
```diff
@@ -616,7 +616,7 @@ def sample_posterior_predictive(self, X_pred, extend_idata, combined, **kwargs):
             post_pred = pm.sample_posterior_predictive(self.idata, **kwargs)
             if extend_idata:
-                self.idata.extend(post_pred)
+                self.idata.extend(post_pred, join="right")
```
Same here. Maybe I missed a spot, so let me know

ricardoV94 replied:
I saw those. What are you referring to? I mentioned `fit` will still override the previous result, which is not something that needs to be fixed in this PR.

The test has the wrong assert because it's checking that the idata matches the first call (not the second). The test is also insufficient, because it doesn't raise (I think you didn't change it when you switched from "left" to "right"

wd60622 replied:
I modified the tests in order to check for the second sample stays. Please take a look

ricardoV94 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -814,3 +853,73 @@ def plot_channel_contributions_grid(
+    def sample_posterior_predictive(
+        self,
+        X_pred,
+        extend_idata,
```
Default missing? Docstrings say it defaults to False, All other methods seem to have a default.

I am not sure what the default should be, probably True?

wd60622 replied:
I've changed all but the sample_prior_predictive (from model_builder) to default to True. Changed in the docstring as well

wd60622 commented:
Hi @ricardoV94. Thanks for the feedback. I just went through all of it and made some changes. 

For the xarray tests, I changed to used check for the second value and am using xr.testing module now for more robust checks.

juanitorduz commented:
LGTM 🚀 ! Let's wait for @ricardoV94 's feedback :)

ricardoV94 commented:
@wd60622 everything looks kosher, but there is still this open discussion: https://github.com/pymc-labs/pymc-marketing/pull/482#discussion_r1457076712

Given how all sampling/fitting methods require x/y anyway I changed my mind and I don't think the `_reset_data` call is needed after all? The less magic probably the better

wd60622 commented:
> @wd60622 everything looks kosher, but there is still this open discussion: [#482 (comment)](https://github.com/pymc-labs/pymc-marketing/pull/482#discussion_r1457076712)
> 
> Given how all sampling/fitting methods require x/y anyway I changed my mind and I don't think the `_reset_data` call is needed after all? The less magic probably the better|

Perfect. I've changed that back so that the data will not be reset

ricardoV94 commented on tests/mmm/test_delayed_saturated_mmm.py:
```diff
@@ -580,6 +609,122 @@ def test_model_config(
+        new_dates,
+    )
+
+    # Data has not been set back
```
This check should now also be removed. Having a test suggests it's a feature when it's just what it is. This could be confusing for someone refactoring the code in the future because it suggests it's something that needs to be satisfied.

juanitorduz commented:
Are we ready to merge? 🤞

juanitorduz commented:
It's green! 💚

ricardoV94 commented:
Merged, any notebooks that need updating?

juanitorduz commented:
> Merged, any notebooks that need updating?

I ran the notebook in my first review and it worked perfectly. I can check tomorrow (@wd60622 do we need an update?).

wd60622 commented:
Firstly, thank you both for the reviews! 

> > Merged, any notebooks that need updating?
> 
> I ran the notebook in my first review and it worked perfectly. I can check tomorrow (@wd60622 do we need an update?).

I haven't touched the notebooks. I would like to add an example in docstring / notebook with https://github.com/pymc-labs/pymc-marketing/issues/494
I will check get to that on the weekend
